### ISSUE TITLE
Typings for Command.add are missing types for the subject

### ISSUE THREAD

**Issue by SimeonC** [cypress-io/cypress#18879]
### Current behavior

A simple custom command that relies on prevSubject is not correctly typed.

### Desired behavior

The "previous subject" argument should be correctly added to the command type and error if not passed.

### Test code to reproduce

```ts
declare global {
  namespace Cypress {
    interface Chainable {
      confirmClick(): void;
    }
  }
}

Cypress.Commands.add('confirmClick', { prevSubject: true }, (confirmButton) => {
  confirmButton.click();
  cy.wrap(confirmButton)
    .should('have.attr', 'data-state', 'confirming')
    .click({ force: true });
});

```

### Cypress Version

9.0.0

**Comment by omersomuncu** [comment]
For me "confirmClick" is not. Neither add nor overwrite is working at the moment.

**Comment by SimeonC** [comment]
I have done some attempts at fixing this for `add` (though I think the fix only works in TS 4.4) but overwrite is significantly more difficult as we don't have Options to switch on.

**Comment by SimeonC** [comment]
My current workaround for this locally is to use a wrapper function as below and change


```diff
- Cypress.Commands.add('confirmClick', { prevSubject: true }, (confirmButton) => {
+ addCypressCommand('confirmClick', { prevSubject: true }, (confirmButton) => {
```

```ts
type AddSubjectArgument<
  TPrevSubject extends Cypress.CommandOptions['prevSubject'],
  TFunction extends (...args: any) => any
> = (
  ...args: [
    TPrevSubject extends true
      ? JQuery | HTMLElement
      : undefined | JQuery | HTMLElement,
    ...Parameters<TFunction>
  ]
) => ReturnType<TFunction>;

function addCypressCommand<
  T extends keyof Cypress.Chainable,
  TOptions extends Cypress.CommandOptions
>(
  name: T,
  options: TOptions,
  fn: Cypress.Chainable[T] extends (...args: any) => any
    ? TOptions['prevSubject'] extends false
      ? Cypress.Chainable[T]
      : AddSubjectArgument<TOptions['prevSubject'], Cypress.Chainable[T]>
    : never
): void {
  Cypress.Commands.add<T>(name, options, fn as Cypress.Chainable[T]);
}
```

**Comment by kubijo** [comment]
> My current workaround for this locally is to use a wrapper function as below and change
> 
> ```diff
> - Cypress.Commands.add('confirmClick', { prevSubject: true }, (confirmButton) => {
> + addCypressCommand('confirmClick', { prevSubject: true }, (confirmButton) => {
> ```
> …

That is not the whole story, though… `{ prevSubject: 'optional' }` also adds subject field.
My solution that was working until 9.0 was to have two types that type-cast the functions, and I was choosing manually 

```typescript
Cypress.Commands.add('formInput', formCommandOptions, (…) as CommandWithContext<'formInput'>;
Cypress.Commands.add('apiIntercept', (…) as Command<'apiIntercept'>);
```

where these are the two types

```typescript
export type Command<T extends keyof CustomCommands> = CustomCommands[T];
export type CommandWithContext<T extends keyof CustomCommands> = {
    (context: Cypress.Chainable | null, ...args: Parameters<CustomCommands[T]>): ReturnType<CustomCommands[T]>;
};
```

…that is much the same idea as your internal solution, just manually forced…

--------------------------

That being said, I have been able to get it working somewhat by re-typing the `Cypress.Commands` to:

```typescript
interface Cypress {
  Commands: {
    add<T extends keyof Chainable>(name: T, fn: Chainable[T]): void
    add<T extends keyof Chainable, Options extends CommandOptions>(
        name: T,
        options: Options,
        fn: Options['prevSubject'] extends string | true
            ? (ctx: Cypress.Chainable | null, ...args: Parameters<Chainable[T]>) => ReturnType<Chainable[T]>
            : Chainable[T]
    ): void
    overwrite<T extends keyof Chainable>(name: T, fn: Chainable[T]): void
  }
}
```

**Comment by kubijo** [comment]
You could possibly extend the conditional to check for the `prevSubject` field value more strictly, but I believe this gets the idea accross

**Comment by SimeonC** [comment]
My workaround does handle all values of prevSubject, just in 2 places rather than one.
```ts
  fn: Cypress.Chainable[T] extends (...args: any) => any
    ? TOptions['prevSubject'] extends false
      ? Cypress.Chainable[T]
      : AddSubjectArgument<TOptions['prevSubject'], Cypress.Chainable[T]>
    : never
```
At this place we check if prevSubject is false, which means that the subject should not be added - hence returning `Cypress.Chainable[T]` otherwise we pass to the `AddSubjectArgument` type.

Key is this part of the type at this point, our type of TPrevSubject is `true | string | string[]` (OK, not quite string cause it's 'internal', 'window' etc). So the ternary correctly handles between being true and the value cannot be undefined, or the other values where it can be undefined.

```ts
...args: [
    TPrevSubject extends true
      ? JQuery | HTMLElement
      : undefined | JQuery | HTMLElement,
    ...Parameters<TFunction>
```

This is the solution I used in the attached PR which passes all the TS 4 tests, the only problem I have is that this only seems to work in TS 4 so the TS 3 tests fail.

**Comment by maciaszczykm** [comment]
We have the same problem, typings are not available for child commands since https://github.com/cypress-io/cypress/pull/17496 was introduced.

**Comment by Threnos** [comment]
Bump

**Comment by FrancisBourgault** [comment]
Same issue here.

All my commands already have typescript declaration, but now it breaks because the prevSubject parameter isn't present in the declaration (and it shouldn't be, since it's not directly provided by the user)

**Comment by SimeonC** [comment]
This will now be resolved in #19003
